uefi/src/proto/shell/mod.rs

```diff
@@ -0,0 +1,584 @@
+        out_file_list: *mut *mut ShellFileInfo,
+    ) -> Status,
+    find_files_in_dir: extern "efiapi" fn(
+        file_dir_handle: ShellFileHandle,
+        out_file_list: *mut *mut ShellFileInfo,
+    ) -> Status,
+    get_file_size: extern "efiapi" fn(file_handle: ShellFileHandle, size: *mut u64) -> Status,
+
+    open_root: usize,
+    open_root_by_handle: usize,
+
+    execution_break: Event,
+
+    major_version: u32,
+    minor_version: u32,
+    register_guid_name: usize,
+    get_guid_name: usize,
+    get_guid_from_name: usize,
+    get_env_ex: usize,
+}
+
+impl core::fmt::Debug for Shell {
+    fn fmt(&self, _f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        todo!()
+    }
+}
+
+/// Enum describing output options for get_env()
+///
+/// `EnvOutput::Val` - Value for a given variable
+/// `EnvOutput::VarVec` - Vec of current environment variables
+#[derive(Debug)]
+pub enum EnvOutput<'a> {
+    /// Value for a given variable
+    Val(&'a CStr16),
+    /// Vec of current environment variable names
+    Vec(Vec<&'a CStr16>),
+}
+
+impl<'a> EnvOutput<'a> {
+    /// Extracts the env var value from EnvOutput
+    #[must_use]
+    pub fn val(self) -> Option<&'a CStr16> {
+        match self {
+            EnvOutput::Val(v) => Some(v),
+            _ => None,
+        }
+    }
+
+    /// Extracts the vector of variable names from EnvOutput
+    #[must_use]
+    pub fn vec(self) -> Option<Vec<&'a CStr16>> {
+        match self {
+            EnvOutput::Vec(v) => Some(v),
+            _ => None,
+        }
+    }
+}
+
+impl Shell {
+    /// TODO
+    pub fn execute(
+        &self,
+        parent_image: Handle,
+        command_line: &CStr16,
+        environment: &[&CStr16],
+    ) -> Result<Status> {
+        let mut out_status: MaybeUninit<Status> = MaybeUninit::uninit();
+        // We have to do this in two parts, an `as` cast straight to *const *const CStr16 doesn't compile
+        // let environment = environment.as_ptr();
+        // let environment = environment.cast::<*const CStr16>();
+
+        let cl_ptr = command_line.as_ptr();
+        unsafe {
+            let env_ptr: *const *const Char16 = (*environment.as_ptr()).as_ptr().cast();
+
+            (self.execute)(&parent_image, cl_ptr, env_ptr, out_status.as_mut_ptr())
+                .to_result_with_val(|| out_status.assume_init())
+        }
+    }
+
+    /// Gets the environment variable or list of environment variables
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - The environment variable name of which to retrieve the
+    ///   value
+    ///   If None, will return all defined shell environment
+    ///   variables
+    ///
+    /// # Returns
+    ///
+    /// * `Some(env_value)` - Value of the environment variable
+    /// * `Some(Vec<env_names>)` - Vector of environment variable names
+    /// * `None` - Environment variable doesn't exist
+    #[must_use]
+    pub fn get_env<'a>(&'a self, name: Option<&CStr16>) -> Option<EnvOutput<'a>> {
```

Comment:
I'd much rather have the high-level API to not replicate the low-level interface here. Instead, we can use the low-level interface to create an API similar to:
- [`std::env::var`](https://doc.rust-lang.org/std/env/fn.var.html)
- [`std::env::vars`](https://doc.rust-lang.org/std/env/fn.vars.html)

Then we also don't need the `EnvOutput` type effectively acting as a `Either` type.